nixos/nextcloud: Add option to pre-compress assets

Mynacol opened the pull request:
###### Description of changes

The schema is similar to Mastodon's packaging script, gnerating gzip and brotli compressed files ahead-of-time.

The find command only selects those folders that are publicly viewable according to the nginx config, and only compresses file types that are allowed in the nginx config.
The index.html and robots.txt files are too small to be relevant.

The gzip files are used through the default nginx config, brotli requires the administrators to set `services.nginx.recommendedBrotliSettings = true`.

###### Space changes
560MB /nix/store/i6b1aqrbjgvx6jc8psv1kcyhkayqg8xx-nextcloud-27.0.0 (previous)
683MB /nix/store/h2ai1zkdqp280xv4wp7xrfn4ax86l9xf-nextcloud-27.0.0 (new)

###### Bandwidth savings
I observed the login page with Firefox. The server has `services.nginx.recommendedGzipSettings` and `services.nginx.recommendedBrotliSettings` enabled.
_16.53MB uncompressed (according to Firefox dev tools)_
3.94MB on-the-fly compressed (old)
1.64MB ahead-of-time compressed (new)
That's a 2.4x compression improvement!

###### Compute savings
No measurements for CPU compute savings on the server side, but we avoid compressing the same files over and over again. Helps the environment :earth_africa: 

Brotli and gzip should be equally fast at decompression irrespective of compression level.

###### Build time increase
The build time increases by 15-23 minutes. Most of this can be attributed to zopfli. Only brotli (and standard gzip) would take more like 4 minutes.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`): **tested nextcloud26 in production**
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ajs124 commented:
This will break the integrity check, no?

Ma27 commented:
I'd strongly prefer to have the compression artifacts moved to another output (i.e. to make this opt-in).

> Most of this can be attributed to zopfli

Please elaborate why we also create zopfli artifacts in addition to gzip and brotli.

Ma27 commented:
> Should we add another option, and then using symlinkJoin to create a combined document root?

Rough idea: `enableCompressedAssets`-option. Not sure if that works, but ideally one could simply do a `try_files` then which looks in `nextcloud.compressed` first and then in `nextcloud.out`.
Otherwise, you may wanna take a look at how `gitea.data-compressed` works and then adapt it accordingly here.

Ma27 commented:
Random thought: if we only use the `symlinkJoin` for assets, but the "correct" package for the PHP stuff then we could "fix" the integrity check errors, correct?

SuperSandro2000 commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -28,6 +28,26 @@ let
       inherit packages;
     };
 
+    outputs = [ "out" "compressed" ];
+
+    nativeBuildInputs = [ zopfli brotli ];
+
+    buildPhase = ''
```
Please move this into passthru. Someone not wanting compressed data should be required to run this for potentially minutes.

SuperSandro2000 commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -28,6 +28,26 @@ let
+          case "$f" in
+          *.css | *.js | *.json | *.svg | *.ico | *.txt | *.md | *.xml | *.html | *.ttf | *.otf | *.eot)
+            mkdir -p "$compressed/''${f%/*}"
+            zopfli --gzip -c "$f" > "$compressed/$f.gz"
+            brotli --best --keep --output="$compressed/$f.br" "$f"
```
both command support putting the file next to the existing file. please see https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/gitea/default.nix#L76-L79

SuperSandro2000 commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -28,6 +28,26 @@ let
+    nativeBuildInputs = [ zopfli brotli ];
+
+    buildPhase = ''
+      # Create missing static gzip and brotli files
```
They are not "missing" 

```suggestion
      # Create static compressed gzip and brotli files for reverse proxies like nginx
```

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -218,6 +218,12 @@ in {
+    enableCompressedAssets = mkOption {
+      type = types.bool;
+      default = true;
+      description = lib.mdDoc "Whether to use gzip and brotli pre-compressed artifacts for lower CPU usage and transfer size but bigger storage size.";
```
```suggestion
      description = lib.mdDoc "Whether to use gzip and brotli pre-compressed assets for lower CPU usage in production and less transfer size.";
```
the last is only true after moving the compression into passthru

SuperSandro2000 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -744,8 +750,16 @@ in {
+                paths = [ pkg.out pkg.compressed ];
+              }
+            else pkg
+          );
```
We don't need to symlink join, just replace the package with the compressed variant. see https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/gitea/default.nix#L69-L83 for an example

SuperSandro2000 commented:
> Please elaborate why we also create zopfli artifacts in addition to gzip and brotli.

Yes, it will because it complains about the new files.

(I just wanted to attach a screenshot but somehow the error is now not appearing. Not sure why but that already happened in the past for no reason and after some time it appeared again.)

SuperSandro2000 commented:
you probably want to implement the change here https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/nextcloud/default.nix#L14

Ma27 commented:
I think this should use https://github.com/NixOS/nixpkgs/pull/292324 as soon as it's ready.

SuperSandro2000 commented:
Linked PR got merged

Ma27 commented:
I must say a of this back and forth and the fact that this will add more complexity to Nextcloud and others makes me heavily question the gain of pre-compressed artifacts over letting nginx do it.

Mynacol committed:
nixos/nextcloud: Add option to pre-compress assets

This adds an option which pre-compresses assets of the nextcloud webroot
at system build time using `compressDrvWeb`.

The nextcloud nginx config unconditionally activates gzip compression to
significantly reduce the transmission size of these assets. By
pre-compressing them, even better compression factors can be reached
while avoiding the constant on-the-fly compression of the same files.
The advantages are even bigger when using sendfile on either an
unencrypted HTTP connection or in combination with KTLS by avoiding a
copy into the userspace.
The trade-off is an increased storage size holding these pre-compressed
artifacts.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -444,6 +447,12 @@ in
 
+    enableCompressedAssets = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
```
Am I getting it right that we're serving storing the same data compressed in _several_ formats now by default?

Ma27 reviewed (commented):
Did anybody actually measure if the improvements are noticeable?

CCing the actual Nextcloud maintainers as well: @bachp @britter @provokateurin @dotlambda 

> Would it make sense to add the nextcloud nixos tests to the nixos module's passthru.tests (so ofborg will run them on any change here)?


This is already the case for quite a while though?

Ma27 commented:
> At least transfer size wise is it noticeable. See the original post for (old, but probably comparable) numbers.

First of all, no measurement of the timings, i.e. what the actual impact is when using on-the-fly compression. Then, what are the compression parameters used for on-the-fly vs. build-time compression (and the timing-differences when using actually equivalent configuration)?

Btw, what's the reason for doing this in nixpkgs instead of when building the upstream tarball?

The one thing I'm interested in is that this isn't the default considering the build-time compression speeds this will make testing cycles way more annoying.

I'll defer the decisionmaking to someone else from the maintainer team though.

provokateurin commented:
I'm not really convinced by this change. Browser cache assets and you can also configure nginx to cache assets (but you can easily break your instance on upgrade if your cache invalidation is not working), so the amount of on-the-fly compression should not be high. Even though it's optional, IMO this adds complexity without much benefit. You can also easily have this in your NixOS configuration by overriding the `services.nextcloud.package` option without needing this change in the module. So I'm not in favor of merging this, but it could be added to the docs instead.
